packages/not-next/src/components/Listings.tsx

```diff
@@ -0,0 +1,88 @@
+// SPDX-FileCopyrightText: 2024 Mass Labs
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+import { useEffect, useState } from "react";
+
+import CustomerViewProducts from "./CustomerViewListings.tsx";
+import { useKeycard } from "../hooks/useKeycard.ts";
+import { useClientWithStateManager } from "../hooks/useClientWithStateManager.ts";
+import MerchantViewProducts from "./merchants/listings/MerchantViewListings.tsx";
+
+export default function Listings() {
+  const { clientStateManager } = useClientWithStateManager();
+  const [keycard] = useKeycard();
+  const [products, setProducts] = useState(new Map());
+
+  async function getAllListings() {
+    const listings = new Map();
+    for await (
+      const [
+        id,
+        item,
+      ] of clientStateManager!.stateManager.listings.iterator()
+    ) {
+      listings.set(id, item);
+    }
+    return listings;
+  }
+
+  useEffect(() => {
+    async function onCreateEvent() {
+      const l = new Map();
+      for await (
+        const [
+          id,
+          item,
+        ] of clientStateManager!.stateManager.listings.iterator()
+      ) {
+        l.set(id, item);
+      }
+      setProducts(l);
+    }
+    async function onUpdateEvent() {
+      const l = new Map();
+      for await (
+        const [
+          id,
+          item,
+        ] of clientStateManager!.stateManager.listings.iterator()
+      ) {
+        l.set(id, item);
+      }
```

Comment:
this code is repeated several times, write a function `AsyncIteratorToMap<K,V>(iterator: AsyncIterator): Map<K,V>` to replace the duplicate code. Or use a [library like](https://github.com/ReactiveX/IxJS?tab=readme-ov-file#asynciterable)